commit 72be9d2df5d6b4da21a9b80811035ab26fbabe9e
Author: Dmitry Grinberg <dmitrygr@gmail.com>
Date:   Thu Dec 30 21:21:56 2021 -0800

    do not build r4000 TLBEX code for r3000 targets and vice-versa

diff --git a/arch/mips/mm/Makefile b/arch/mips/mm/Makefile
index b4c64bd3..c0ddc311 100644
--- a/arch/mips/mm/Makefile
+++ b/arch/mips/mm/Makefile
@@ -12,6 +12,8 @@ else
 obj-y				+= uasm-mips.o
 endif
 
+ccflags-y			:= -Wno-unused-function
+
 obj-$(CONFIG_32BIT)		+= ioremap.o pgtable-32.o
 obj-$(CONFIG_64BIT)		+= pgtable-64.o
 obj-$(CONFIG_HIGHMEM)		+= highmem.o
diff --git a/arch/mips/mm/tlbex.c b/arch/mips/mm/tlbex.c
index 448b4aab..8ae7deed 100644
--- a/arch/mips/mm/tlbex.c
+++ b/arch/mips/mm/tlbex.c
@@ -1717,6 +1717,9 @@ build_pte_modifiable(u32 **p, struct uasm_reloc **r,
 	}
 }
 
+
+#ifdef CONFIG_CPU_R3000
+
 #ifndef CONFIG_MIPS_PGD_C0_CONTEXT
 
 
@@ -1876,6 +1879,8 @@ static void build_r3000_tlb_modify_handler(void)
 }
 #endif /* CONFIG_MIPS_PGD_C0_CONTEXT */
 
+#else // CONFIG_CPU_R3000
+
 /*
  * R4000 style TLB load/store/modify handlers.
  */
@@ -2235,6 +2240,8 @@ static void build_r4000_tlb_modify_handler(void)
 	dump_handler("r4000_tlb_modify", handle_tlbm, handle_tlbm_size);
 }
 
+#endif // CONFIG_CPU_R3000
+
 static void flush_tlb_handlers(void)
 {
 	local_flush_icache_range((unsigned long)handle_tlbl,
@@ -2418,6 +2425,9 @@ void build_tlb_refill_handler(void)
 #endif
 
 	switch (current_cpu_type()) {
+
+#ifdef CONFIG_CPU_R3000
+
 	case CPU_R2000:
 	case CPU_R3000:
 	case CPU_R3000A:
@@ -2442,6 +2452,7 @@ void build_tlb_refill_handler(void)
 		panic("No R3000 TLB refill handler");
 #endif
 		break;
+#endif // CONFIG_CPU_R3000
 
 	case CPU_R6000:
 	case CPU_R6000A:
@@ -2453,6 +2464,9 @@ void build_tlb_refill_handler(void)
 		break;
 
 	default:
+#ifdef CONFIG_CPU_R3000
+		panic("No R4000 TLB refill handler in R3000 kernel");
+#else
 		if (!run_once) {
 			scratch_reg = allocate_kscratch();
 			build_setup_pgd();
@@ -2470,5 +2484,6 @@ void build_tlb_refill_handler(void)
 			config_xpa_params();
 		if (cpu_has_htw)
 			config_htw_params();
+#endif
 	}
 }
